ampd/src/event_sub/stream.rs

```diff
@@ -0,0 +1,479 @@
+    retry_policy: RetryPolicy,
+) -> Result<Vec<Event>>
+where
+    T: TmClient,
+{
+    match block_height {
+        Ok(block_height) => {
+            with_retry(|| block_events(tm_client, block_height), retry_policy).await
+        }
+        Err(err) => Err(err),
+    }
+}
+
+async fn block_events<T>(tm_client: &T, block_height: block::Height) -> Result<Vec<Event>>
+where
+    T: TmClient,
+{
+    let block_results =
+        tm_client
+            .block_results(block_height)
+            .await
+            .change_context(Error::BlockResultsQuery {
+                block: block_height,
+            })?;
+
+    let begin_block_events = block_results.begin_block_events.into_iter().flatten();
+    let tx_events = block_results
+        .txs_results
+        .into_iter()
+        .flatten()
+        .flat_map(|tx| tx.events);
+    let end_block_events = block_results.end_block_events.into_iter().flatten();
+
+    let events = begin_block_events
+        .chain(tx_events)
+        .chain(end_block_events)
+        .map(|event| {
+            Event::try_from(event).change_context(Error::EventDecoding {
+                block: block_height,
+            })
+        })
+        .collect::<Result<Vec<_>>>()?;
+
+    Ok(iter::once(Event::BlockBegin(block_height))
+        .chain(events)
+        .chain(iter::once(Event::BlockEnd(block_height)))
+        .collect())
+}
+
+#[derive(Copy, Clone)]
+struct BlockState {
+    current: block::Height,
+    latest: block::Height,
+}
+
+impl BlockState {
+    async fn update<T>(mut self, tm_client: &T, interval: &mut Interval) -> Result<Self>
+    where
+        T: TmClient,
+    {
+        while self.current > self.latest {
+            self.latest = interval
+                .tick()
+                .then(|_| latest_block_height(tm_client))
+                .await?;
+        }
+
+        self.current = self.current.increment();
+
+        Ok(self)
+    }
+
+    fn stream<T>(
+        self,
+        tm_client: &T,
+        interval: Interval,
+        token: CancellationToken,
+    ) -> impl Stream<Item = Result<block::Height>> + '_
+    where
+        T: TmClient,
+    {
+        futures::stream::unfold(
+            (self, tm_client, interval, token),
+            |(block_state, tm_client, mut interval, token)| async move {
+                let current = block_state.current;
+
+                if token.is_cancelled() {
+                    return None;
+                }
+
+                match block_state.update(tm_client, &mut interval).await {
+                    Ok(block_state) => {
+                        Some((Ok(current), (block_state, tm_client, interval, token)))
+                    }
+                    Err(err) => Some((Err(err), (block_state, tm_client, interval, token))),
+                }
+            },
+        )
+    }
+}
+
+async fn latest_block_height<T: TmClient>(tm_client: &T) -> Result<block::Height> {
+    tm_client
+        .latest_block()
+        .await
```

Comment:
does this have a timeout?

Reply:
Not sure, It probably has a default one that's used by the underlying library. We can do a separate PR to apply a timeout ourselves.